debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerSnapshotChangeEventSource.java

```diff
@@ -257,40 +257,37 @@ protected void complete(SnapshotContext<SqlServerPartition, SqlServerOffsetConte
                                                  TableId tableId) {
-        String modifiedColumns = checkExcludedColumns(tableId);
-        return Optional.of(String.format("SELECT %s FROM [%s].[%s]", modifiedColumns, tableId.schema(), tableId.table()));
+        String snapshotSelectColumns = getSnapshotSelectColumns(tableId);
+        return Optional.of(String.format("SELECT %s FROM [%s].[%s]", snapshotSelectColumns, tableId.schema(), tableId.table()));
     }
 
     @Override
-    protected String enhanceOverriddenSelect(RelationalSnapshotContext<SqlServerPartition, SqlServerOffsetContext> snapshotContext,
-                                             String overriddenSelect, TableId tableId) {
-        String modifiedColumns = checkExcludedColumns(tableId);
-        return overriddenSelect.replaceAll("\\*", modifiedColumns);
-    }
-
-    private String checkExcludedColumns(TableId tableId) {
+    protected String getSnapshotSelectColumns(TableId tableId) {
         Table table = sqlServerDatabaseSchema.tableFor(tableId);
-        List<String> columnNames = table.retrieveColumnNames().stream()
-                .filter(columnName -> filterChangeTableColumns(tableId, columnName))
-                .filter(columnName -> connectorConfig.getColumnFilter().matches(tableId.catalog(), tableId.schema(), tableId.table(), columnName))
-                .collect(Collectors.toList());
+        return getPreparedColumnNames(table, tableId).stream().collect(Collectors.joining(","));
 
-        if (columnNames.isEmpty()) {
-            throw new IllegalArgumentException("Filtered column list for table " + tableId + " is empty");
+    }
+
+    @Override
+    protected String resolveColumnName(TableId tableId, String columnName) {
+        StringBuilder sb = new StringBuilder();
+        if (filterChangeTableColumns(tableId, columnName)) {
+            if (!columnName.contains(tableId.table())) {
+                sb.append("[").append(tableId.table()).append("]")
+                        .append(".[").append(columnName).append("]");
+            }
+            else {
+                sb.append("[").append(columnName).append("]");
+            }
         }
+        return sb.toString();
+    }
 
-        return columnNames.stream()
-                .map(columnName -> {
-                    StringBuilder sb = new StringBuilder();
-                    if (!columnName.contains(tableId.table())) {
-                        sb.append("[").append(tableId.table()).append("]")
-                                .append(".[").append(columnName).append("]");
-                    }
-                    else {
-                        sb.append("[").append(columnName).append("]");
-                    }
-                    return sb.toString();
-                }).collect(Collectors.joining(","));
+    @Override
+    protected String enhanceOverriddenSelect(RelationalSnapshotContext<SqlServerPartition, SqlServerOffsetContext> snapshotContext,
```

Comment:
@gunnarmorling The overridden select part of SQL Server would be removed as well? As this was initially in the exclusion logic and exists in the `master`.

Reply:
Ah, that's a great point. Thanks for calling this out. So my general thinking is that the replacement logic here is very brittle and it should not be applied to custom select statements (e.g. there could be a `*` in a sub-select in the WHERE clause). That's why I removed this (see https://issues.redhat.com/browse/DBZ-3766 for a much more reliable approach for this which we should adopt eventually).

But you have a good point that we probably should leave this capability for SQL Server for the time being. So you'd probably have to make `getPreparedColumnNames()` in the base class protected so we can use it here.